src/mlpack/methods/bias_svd/bias_svd_function_impl.hpp

```diff
@@ -0,0 +1,411 @@
+/**
+ * @file bias_svd_function_impl.hpp
+ * @author Siddharth Agrawal
+ * @author Wenhao Huang
+ *
+ * An implementation of the BiasSVDFunction class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_BIAS_SVD_BIAS_FUNCTION_SVD_IMPL_HPP
+#define MLPACK_METHODS_BIAS_SVD_BIAS_FUNCTION_SVD_IMPL_HPP
+
+#include "bias_svd_function.hpp"
+#include <mlpack/core/math/make_alias.hpp>
+
+namespace mlpack {
+namespace svd {
+
+template <typename MatType>
+BiasSVDFunction<MatType>::BiasSVDFunction(const MatType& data,
+                                          const size_t rank,
+                                          const double lambda) :
+    data(math::MakeAlias(const_cast<MatType&>(data), false)),
+    rank(rank),
+    lambda(lambda)
+{
+  // Number of users and items in the data.
+  numUsers = max(data.row(0)) + 1;
+  numItems = max(data.row(1)) + 1;
+
+  // Initialize the parameters.
+  // The last row in initialPoint saves use/item bias.
+  initialPoint.randu(rank + 1, numUsers + numItems);
+}
+
+template<typename MatType>
+void BiasSVDFunction<MatType>::Shuffle()
+{
+  data = data.cols(arma::shuffle(arma::linspace<arma::uvec>(0, data.n_cols - 1,
+      data.n_cols)));
+}
+
+template <typename MatType>
+double BiasSVDFunction<MatType>::Evaluate(const arma::mat& parameters) const
+{
```

Comment:
Actually, we've got a lot of almost duplicated code such as `RegularizedSVD`, `BiasSVD` and `SVDPlusPlus`. Unfortunately, I don't see a proper way to merge them. However we could reduce some code as follows. I think we could replace the whole body of this function by
```
{
  return Evaluate(parameters, 0, data.n_cols);
}
```
What do you think?

Reply:
Yes I also noticed the problem but I don't have any thought on how to properly merge the duplicated code either. For the `Evaluate(parameters)` function you pointed out, yes I think we can do that.